mojo/core/broker_castanets.cc

```diff
@@ -353,13 +392,65 @@ void BrokerCastanets::SyncSharedBufferImpl(const base::UnguessableToken& guid,
 
 void BrokerCastanets::SyncSharedBufferImpl2d(const base::UnguessableToken& guid,
                                              uint8_t* memory,
-                                             size_t offset,
-                                             size_t sync_size,
                                              size_t mapped_size,
                                              size_t width,
+                                             size_t height,
+                                             size_t bytes_per_pixel,
+                                             size_t offset,
                                              size_t stride,
+                                             BrokerCompressionMode compression_mode,
                                              bool write_lock) {
+  size_t sync_size = width * height * bytes_per_pixel;
   CHECK_GE(mapped_size, offset + sync_size);
+  size_t put_bytes = 0;
+  size_t put_offset = offset;
+  uint32_t buffer_stride = width * bytes_per_pixel;
+  uint8_t* start_ptr = nullptr;
+
+  std::vector<uint8_t> bytes(sync_size);
+
+  uint8_t* compressed_data = nullptr;
+  size_t original_sync_size = sync_size;
+
+  switch (compression_mode) {
+    case BrokerCompressionMode::ZLIB: {
+      LOG(ERROR) << "Unsupported compression mode";
+      break;
+    }
+    case BrokerCompressionMode::WEBP: {
+      size_t webp_stride = stride ? stride : buffer_stride;
+      start_ptr = memory + offset;
+      size_t size;
+      size = WebPEncodeLosslessRGBA(start_ptr, width,
+          height, webp_stride, &compressed_data);
```

Comment:
Needs WebPFree() for |compressed_data| to reclaim memory?

Reply:
I have added it in Line 474.  It seems to be ok ?
 if (compression_mode == BrokerCompressionMode::WEBP && compressed_data)
    WebPFree (compressed_data);

Reply:
Looks good:)